### PR TITLE
Create packages only for packaged builds, and publish them when created.

### PR DESCRIPTION
**Pull request by gspencergoog**
This changes the publishing of archives so that it happens on the chrome_infra bots when they build a packaged branch instead of as part of the dev_roll process.

It uses the tagged version in the branch, and leaves the git repo that it clones checked out on the branch and hash used to build the package.

It updates metadata located at gs://flutter_infra/releases/releases_<platform>.json (where <platform> is one of macos, linux, or windows) once published, since it would be complex to do the proper locking to keep them all in one shared .json file safely.

A separate [change to the chrome_infra bots](https://chromium-review.googlesource.com/c/chromium/tools/build/+/902823) was made to instruct them to build packaged for the dev, beta, and release branches (but not master anymore).

### CODE REVIEW

File: dev/bots/prepare_package.dart [hunk 1]
```diff
@@ -20,72 +20,220 @@ const String MINGIT_FOR_WINDOWS_URL = 'https://storage.googleapis.com/flutter_in
...
+enum Branch { dev, beta, stable }
+
+String getBranchName(Branch branch) {
+  switch (branch) {
+    case Branch.beta:
+      return 'beta';
+    case Branch.dev:
+      return 'dev';
+    case Branch.stable:
+      return 'stable';
```
Comment (Hixie) [inline dev/bots/prepare_package.dart]: stable -> release

**Comment by Hixie** [comment]
cc @tvolkert for review

File: dev/bots/prepare_package.dart (hunk 1 above)
Comment (gspencergoog) [reply dev/bots/prepare_package.dart]: Fixed.

File: dev/bots/prepare_package.dart [hunk 2]
```diff
@@ -20,72 +20,220 @@ const String MINGIT_FOR_WINDOWS_URL = 'https://storage.googleapis.com/flutter_in
...
-  String message = '';
-  int exitCode = 0;
+  final String message;
+  final ProcessResult result;
+  int get exitCode => result.exitCode ?? -1;
 
   @override
-  String toString() => message;
+  String toString() {
+    String output = 'ArchivePublisherException';
```
Comment (tvolkert) [inline dev/bots/prepare_package.dart]: ArchivePublisherException?

File: dev/bots/prepare_package.dart [hunk 3]
```diff
@@ -20,72 +20,220 @@ const String MINGIT_FOR_WINDOWS_URL = 'https://storage.googleapis.com/flutter_in
...
+    default:
+      throw new ArgumentError('Invalid branch name.');
+  }
+}
+
+/// A helper class for classes that want to run a process, have the stderr and
+/// stdout reported as the process runs, and capture the stdout.
+class ProcessRunner {
+  ProcessRunner({
+    ProcessManager processManager,
```
Comment (tvolkert) [inline dev/bots/prepare_package.dart]: `this.processManager: const LocalProcessManager()`

File: dev/bots/prepare_package.dart [hunk 4]
```diff
@@ -20,72 +20,220 @@ const String MINGIT_FOR_WINDOWS_URL = 'https://storage.googleapis.com/flutter_in
...
+    this.subprocessOutput: true,
+    this.defaultWorkingDirectory,
+  }) : processManager = processManager ?? const LocalProcessManager() {
+    environment = new Map<String, String>.from(Platform.environment);
```
Comment (tvolkert) [inline dev/bots/prepare_package.dart]: This makes the test sensitive to the real environment.  Use package:platform's `Platform` instead as an argument here, defaulting to `const LocalPlatform()`

File: dev/bots/prepare_package.dart [hunk 5]
```diff
@@ -20,72 +20,220 @@ const String MINGIT_FOR_WINDOWS_URL = 'https://storage.googleapis.com/flutter_in
...
+      throw new ArgumentError('Invalid branch name.');
+  }
+}
+
+/// A helper class for classes that want to run a process, have the stderr and
+/// stdout reported as the process runs, and capture the stdout.
+class ProcessRunner {
+  ProcessRunner({
+    ProcessManager processManager,
+    this.subprocessOutput: true,
```
Comment (tvolkert) [inline dev/bots/prepare_package.dart]: It's not clear from the name what this does -- can you document it or rename it to be more clear?

**Comment by tvolkert** [comment]
How are we going to test that the archives we create from this tool are functional?

**Comment by gspencergoog** [comment]
I suppose for testing the archives, we should unpack the newly created archive and check to make sure that 'flutter packages get --offline' works as intended, and that the flutter tool runs and builds without needing a pub server (e.g. set the environment variable for the pub server to something invalid).  We should probably create and build each type of template too, and make sure that the IntelliJ files are created.

I'm a little worried about making sure that this testing happens in a truly "offline" environment.  Since it's impractical to take the test machines off the network, we might have to add an "--offline" to flutter create and other things that use pub so that it doesn't try to go online.

I'd like to write that testing code in another PR, however.

**Review by tvolkert — APPROVED** [review]
LGTM

**Pull request merged**